appservice/src/wizard.ts

```diff
@@ -0,0 +1,197 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { AzureAccountWrapper } from './azureAccountWrapper';
+import { SubscriptionModels } from 'azure-arm-resource';
+import { UserCancelledError, WizardFailedError } from './errors';
+
+export type WizardStatus = 'PromptCompleted' | 'Completed' | 'Faulted' | 'Cancelled';
+
+export abstract class WizardBase {
+    private readonly _steps: WizardStep[] = [];
+    private _result: WizardResult;
+
+    protected constructor(readonly output: vscode.OutputChannel) { }
+
+    protected abstract initSteps();
+
+    async run(promptOnly = false): Promise<WizardResult> {
+        this.initSteps();
+
+        // Go through the prompts...
+        for (var i = 0; i < this.steps.length; i++) {
+            const step = this.steps[i];
+
+            try {
+                await this.steps[i].prompt();
+            } catch (err) {
+                this.onError(err, step);
+            }
+        }
+
+        if (promptOnly) {
+            return {
+                status: 'PromptCompleted',
+                step: this.steps[this.steps.length - 1],
+                error: null
+            };
+        }
+
+        return this.execute();
+    }
+
+    async execute(): Promise<WizardResult> {
+        // Execute each step...
+        this.output.show(true);
+        for (var i = 0; i < this.steps.length; i++) {
+            const step = this.steps[i];
+
+            try {
+                this.beforeExecute(step, i);
```

Comment:
I don't see the value of 'beforeExecute'. Anything run in there could just be run in the first step